### PR TITLE
[Refactor,Feature] Refactor collector shapes and stack_result in sync collector

### PR DESCRIPTION
**Pull request by vmoens**
## Description

Describe your changes in detail.

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [ ] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by vmoens** [comment]
@albertbou92 I also took this opportunity to refactor preemption. Now you can stack (with padding) or cat (with masking) the values coming from the collectors.
I noticed we preempt based on the proportion of collectors that have reached the end, not on the proportion of frames collected. IMO that would be more sensible.
We could use a `mp.Value()` shared between processes that tracks how many frames have been collected, and as soon as this goes above the threshold we break. Would you like to give it a shot once this PR lands? No worry if you don't have the time!
I suspect it can bring some decent speed-up over what we're doing.

In a nutshell:
```python
# in SyncDataCollector.rollout
def rollout(self):
    ...
    if self.interruptor is not None:
        self._collected_frames += frame_count

# in MultiSyncDataCollector.iterator
def iterator(self):
    if self.preemptive_threshold and ...:
        if self._collected_frames > self.frames_per_batch * self.preemptive_threshold:
            break # or similar
```

**Comment by albertbou92** [comment]
> @albertbou92 I also took this opportunity to refactor preemption. Now you can stack (with padding) or cat (with masking) the values coming from the collectors. I noticed we preempt based on the proportion of collectors that have reached the end, not on the proportion of frames collected. IMO that would be more sensible. We could use a `mp.Value()` shared between processes that tracks how many frames have been collected, and as soon as this goes above the threshold we break. Would you like to give it a shot once this PR lands? No worry if you don't have the time! I suspect it can bring some decent speed-up over what we're doing.
> 
> In a nutshell:
> 
> ```python
> # in SyncDataCollector.rollout
> def rollout(self):
>     ...
>     if self.interruptor is not None:
>         self._collected_frames += frame_count
> 
> # in MultiSyncDataCollector.iterator
> def iterator(self):
>     if self.preemptive_threshold and ...:
>         if self._collected_frames > self.frames_per_batch * self.preemptive_threshold:
>             break # or similar
> ```

Yes the preemption was based on the idea that all workers collect a fixed number of frames and can not communicate during collection, but if we can have a shared global value that tracks how many frames that would be much better. Yes, I can give it shot!

**Comment by vmoens** [comment]
@matteobettini This PR addresses an issue that you pointed a long time ago RE the stacking of results in MultiSync
Do you think it solves it in an adequate way?

The idea is to build the collector with `cat_results="stack"` if you want to stack or `cat_results=-1` if you want to cat along the time dimension. `cat_result=None` results in `cat_results=0` + warning since this is the previous behaviour but we want to discourage it. 
We recommend "stack" > -1 > 0

Sorry it took so long to solve this!

**Comment by matteobettini** [comment]
> @matteobettini This PR addresses an issue that you pointed a long time ago RE the stacking of results in MultiSync Do you think it solves it in an adequate way?
> 
> The idea is to build the collector with `cat_results="stack"` if you want to stack or `cat_results=-1` if you want to cat along the time dimension. `cat_result=None` results in `cat_results=0` + warning since this is the previous behaviour but we want to discourage it. We recommend "stack" > -1 > 0
> 
> Sorry it took so long to solve this!

Amamzing!

So `stack_results=True` now is exactly like a single collector with the number of collectors B preappended to the batch size which feels so smooth!

This is what i wanted yes!

File: docs/source/reference/collectors.rst [hunk 1]
```diff
@@ -59,6 +59,49 @@ Besides those compute parameters, users may choose to configure the following pa
...
+the specificities of the collector. The following table summarizes what is to
+be expected when collecting data:
+
+
++--------------------+---------------------+-----------------------------+------------------------------+
+|                    | SyncDataCollector   |MultiSyncDataCollector (n=B) |MultiaSyncDataCollector (n=B) |
++====================+=====================+=============+===============+==============================+
+|   `stack_result`   |          NA         |    `True`   |    `False`    |             NA               |
++--------------------+---------------------+-------------+---------------+------------------------------+
+|     Single env     |         [T]         |   `[B, T]`  |  `[B*(T//B)`  |              [T]             |
```
Comment (matteobettini) [inline docs/source/reference/collectors.rst:76]:
```suggestion
|     Single env     |         [T]         |   `[B, T]`  |  `[B*(T//B)]`  |              [T]             |
```

File: docs/source/reference/collectors.rst [hunk 2]
```diff
@@ -59,6 +59,49 @@ Besides those compute parameters, users may choose to configure the following pa
...
++--------------------+---------------------+-------------+---------------+------------------------------+
+| Batched env (n=P)  |       [P, T]        | `[B, P, T]` |  `[B * P, T]` |            [P, T]            |
```
Comment (matteobettini) [inline docs/source/reference/collectors.rst:78]: `[B * P, T]`, this is given that ` cat_results=0`

my personal preference, to avoid having `cat_results` be both int and str is to have 2 args
- `cat_results` true or false
- `collectors_dim` dimension where to cat or stack

**Review by matteobettini — APPROVED** [review]

File: docs/source/reference/collectors.rst (hunk 2 above)
Comment (vmoens) [reply docs/source/reference/collectors.rst]: but then what is `collector_dim=0` and `cat_results=False`? That should not be allowed. So we need a complicated doc that lists the available configs
If we allow them we also need to test every single combination...

File: docs/source/reference/collectors.rst [hunk 3]
```diff
@@ -59,6 +59,112 @@ Besides those compute parameters, users may choose to configure the following pa
...
+
+
++--------------------+---------------------+--------------------------------------------+------------------------------+
+|                    | SyncDataCollector   |       MultiSyncDataCollector (n=B)         |MultiaSyncDataCollector (n=B) |
++====================+=====================+=============+==============+===============+==============================+
+|   `cat_results`    |          NA         |  `"stack"`  |      `0`     |      `-1`     |             NA               |
++--------------------+---------------------+-------------+--------------+---------------+------------------------------+
+|     Single env     |         [T]         |   `[B, T]`  |  `[B*(T//B)` |  `[B*(T//B)`  |              [T]             |
++--------------------+---------------------+-------------+--------------+---------------+------------------------------+
+| Batched env (n=P)  |       [P, T]        | `[B, P, T]` |  `[B * P, T]`|  `[B, T * P]` |            [P, T]            |
```
[the remark this answers is not in the document]
Comment (vmoens) [reply docs/source/reference/collectors.rst:78]: I don't think so, if you cat along -1 you still have B at the beginning

File: docs/source/reference/collectors.rst (hunk 2 above)
Comment (vmoens) [reply docs/source/reference/collectors.rst]: Note that the goal is to get rid of this arg in v0.6 or v0.7 so adding 2 args is twice as bad as adding 1 given that we want to break things on the long run

File: docs/source/reference/collectors.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (matteobettini) [reply docs/source/reference/collectors.rst:78]: oh i thought B is the dimension that doens't show up when you cat, this is counter intuitive to me

File: docs/source/reference/collectors.rst (hunk 2 above)
Comment (matteobettini) [reply docs/source/reference/collectors.rst]: > collector_dim=0 and cat_results=False

isn't this what happens currently when `cat_results="stack"`?

Anyway yes how you prefer

File: docs/source/reference/collectors.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (matteobettini) [reply docs/source/reference/collectors.rst:78]: in my view of this

- stack -> B will show up (forcefully at 0)
- cat -> B will be cat to the shape index provided

File: docs/source/reference/collectors.rst (hunk 2 above)
Comment (vmoens) [reply docs/source/reference/collectors.rst]: Yes sorry, I meant collector_dim=-1 or 1 or anything that isn't 0

you can't stack along any dim (and should not be able to do so), that's what I mean. Stack should be along dim 0

File: docs/source/reference/collectors.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (vmoens) [reply docs/source/reference/collectors.rst:78]: Ah I think you're right lol 
P is the dim of the env!

File: docs/source/reference/collectors.rst (hunk 2 above)
Comment (matteobettini) [reply docs/source/reference/collectors.rst]: i don't super see this as in the example table we are referring to, in the row for stacking in multisync i can see both

- B, P, T (collector_dim=0 or -3, cat_results=False) (currently the only available option)
- P, B, T (collector_dim=1 or -2, cat_results=False)
- P, T, B (collector_dim=2 or -1, cat_results=False)

Comment (vmoens) [reply docs/source/reference/collectors.rst]: There is no `cat_results=False`, it can only be a string ("stack") or an int
cat_results="stack" => stack B collectors along 0 => [B, P, T]
cat_results=0 => cat B collectors along 0 => [B * P, T]
cat_results=-1 => cat B collectors along -1 => [B, P * T]

Comment (vmoens) [reply docs/source/reference/collectors.rst]: I think I didn't get what you were trying to say though

Comment (matteobettini) [reply docs/source/reference/collectors.rst]: I know, that is the point of this thread, I am discussing why for me it makes sense to separate them.

Comment (vmoens) [reply docs/source/reference/collectors.rst]: > P, B, T (collector_dim=1 or -2, cat_results=False)
> P, T, B (collector_dim=2 or -1, cat_results=False)

I don't think we should support these. If you want that you can transpose your resulting tensordict. Implementing it would require a lot of expensive tests for a feature we will deprecate in 2 releases.
